src/test/stateful_modelinstance_test.cpp

```diff
@@ -174,6 +244,210 @@ TEST_F(StatefulModelInstanceTempDir, modelInstanceFactory) {
     ASSERT_TRUE(typeid(*modelInstance) == typeid(ovms::StatefulModelInstance));
 }
 
+TEST_F(StatefulModelInstanceTempDir, statefulInferMultipleInferences) {
+    ConstructorEnabledModelManager manager;
+    std::unique_ptr<ovms::ModelInstanceUnloadGuard> unload_guard;
+    createConfigFileWithContent(ovmsConfig, configFilePath);
+    auto status = manager.loadConfig(configFilePath);
+    ASSERT_TRUE(status.ok());
+    auto modelInstance = manager.findModelInstance(dummyModelName);
+
+    uint64_t seqId = 1;
+
+    RunStatefulPredicts(modelInstance, modelInput, 100, seqId, nullptr, nullptr);
+    auto stetefulModelInstance = std::static_pointer_cast<ovms::StatefulModelInstance>(modelInstance);
+
+    ASSERT_EQ(stetefulModelInstance->getSequenceManager()->getSequencesCount(), 0);
+}
+
+TEST_F(StatefulModelInstanceTempDir, statefulInferMultipleThreads) {
+    ConstructorEnabledModelManager manager;
+    std::unique_ptr<ovms::ModelInstanceUnloadGuard> unload_guard;
+    createConfigFileWithContent(ovmsConfig, configFilePath);
+    auto status = manager.loadConfig(configFilePath);
+    ASSERT_TRUE(status.ok());
+    auto modelInstance = manager.findModelInstance(dummyModelName);
+
+    uint64_t startingSequenceId = 0;
+    uint16_t numberOfThreads = 100;
+
+    std::vector<std::promise<void>> releaseWaitAfterSequenceStarted1(numberOfThreads), releaseWaitBeforeSequenceFinished1(numberOfThreads);
+    std::vector<std::thread> inferThreads;
+
+    for (auto i = 0u; i < numberOfThreads / 2; ++i) {
+        startingSequenceId++;
+
+        inferThreads.emplace_back(
+            std::thread(
+                [this, &releaseWaitAfterSequenceStarted1, i, startingSequenceId, modelInstance]() {
+                    RunStatefulPredicts(modelInstance, modelInput, 100, startingSequenceId,
+                        std::move(std::make_unique<std::future<void>>(releaseWaitAfterSequenceStarted1[i].get_future())),
+                        nullptr);
+                }));
+
+        startingSequenceId++;
+
+        inferThreads.emplace_back(
+            std::thread(
+                [this, &releaseWaitBeforeSequenceFinished1, i, startingSequenceId, modelInstance]() {
+                    RunStatefulPredicts(modelInstance, modelInput, 100, startingSequenceId,
+                        nullptr,
+                        std::move(std::make_unique<std::future<void>>(releaseWaitBeforeSequenceFinished1[i].get_future())));
+                }));
+    }
```

Comment:
I think it would be clearer if we break it to two loops and introduce new variables like: `numberOfThreadsWaitingOnStart`, `numberOfThreadsWaitingOnEnd`. The we could remove `numberOfThreads / 2` as it's a bit confusing and `startingSequenceId` could be set separately for both loops like:
```

for (uint64_t startingSequenceId = 0, auto i = 0u; i < numberOfThreadsWaitingOnStart; ++i, ++startingSequenceId ) {
 // Do job
}

for (uint64_t startingSequenceId = 0, auto i = 0u; i < numberOfThreadsWaitingOnEnd; ++i, ++startingSequenceId) {
 // Do job
}
```